### PR TITLE
Started python mini course

### PR DESCRIPTION
**Pull request by jerclark**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -150,7 +150,7 @@ Each day I take one subject from the list below, read it cover to cover, take no
 - [x] [The Machine Learning Mastery Method](http://machinelearningmastery.com/machine-learning-mastery-method/)
 - [x] [Machine Learning for Programmers](http://machinelearningmastery.com/machine-learning-for-programmers/)
 - [x] [Applied Machine Learning with Machine Learning Mastery](http://machinelearningmastery.com/start-here/)
-- [ ] [Python Machine Learning Mini-Course](http://machinelearningmastery.com/python-machine-learning-mini-course/)
+- [ ] [(In Progress) Python Machine Learning Mini-Course](http://machinelearningmastery.com/python-machine-learning-mini-course/)
```
Comment (jerclark) [inline README.md]: - Installed python and dependencies on both machines
- began tutorial content from python website. Some interesting notes
 - for loops iterate over sequence objects only
 - slicing works on strings and arrays with [start:end] syntax. No start/end will create a copy
 - "foo" "bar" literally, will return "foobar".

Comment (jerclark) [reply README.md]: Damn - read about numpy at [this link](http://cs231n.github.io/python-numpy-tutorial/) really mathy. Understood most of it, but probably need to go back and review:
* Boolean indexing
* Broadcasting

Comment (jerclark) [reply README.md]: [scratch work in numpy](https://github.com/jerclark/machine-learning-for-software-engineers/blob/python-mini-course/python_work_numpy.txt)

Comment (jerclark) [reply README.md]: matplotlib.pyplot - followed the tutorial here:
http://matplotlib.org/users/pyplot_tutorial.html

import matplotlib.pyplot as plt
plt.plot([1,2],[1,2]) -- will create a line with points (1,1), (2,2)
Can use matlab style line styling
plt.show() will display the plot.
plt.axes() to configure axes
plt.setp(lines) to conifgure lines

use figure() to create a grid of plots.

plt.text(60, .025, r'$\mu=100,\ \sigma=15$') - to display text on the plot at the given location (60, .025)
text() supports TeX expressions.

annotate() will annotate a value on the line.
plt.annotate('local max', xy=(2, 1), xytext=(3, 1.5),
            arrowprops=dict(facecolor='black', shrink=0.05),
            )

Scale can change to log and logit:
plt.xscale(‘log’)

Comment (jerclark) [reply README.md]: Read over the basics of pandas data series and data frames here:
http://pandas.pydata.org/pandas-docs/stable/dsintro.html

Generally good for tabular data.

Comment (jerclark) [reply README.md]: Pandas work [here](https://github.com/jerclark/machine-learning-for-software-engineers/blob/python-mini-course/python_work_pandas.txt)

Comment (jerclark) [reply README.md]: Practice importing data from CSV.

Used the abalone data set from here: http://archive.ics.uci.edu/ml/datasets/Abalone

Imported using:
Python core: csv.reader(pythonfile)
numpy: np.loadtxt('filepath', delimiter=',', etc...)
pandas: pd.read_csv - http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html

Use 'converter' function to convert abalone rings to age by adding 1.5 as advised in the data set site. 

Exercises [here](https://github.com/jerclark/machine-learning-for-software-engineers/blob/python-mini-course/csv-load-exercises.txt)

Comment (jerclark) [reply README.md]: Worked on descriptive stats with pandas data frames:

head()shows first few rows
shape() shows the dimenions and scale
dtypes property describes data types of each column
describe() shows mean, stdev and distributions (25, 50, 75th percentiles)
corr() shows pairwise correlation (using pearson correlation by default)

I wrote a pearson correlation function to get a little practice in python and to bone up on these types of stats usages:
```
def r(data, colx, coly):
        meanx = data.describe()[colx]["mean"]
	meany = data.describe()[coly]["mean"]
	stdx = data.describe()[colx]["std"]
	stdy = data.describe()[coly]["std"]
	n = data.shape[0]
	s = 0
	for ix, row in data.iterrows():
		cx = ((row[colx] - meanx) / stdx)
		cy = ((row[coly] - meany) / stdy)
		s = s + (cx) * (cy)
	r = s * (float(1)/(n-1))
	return r
```

pariwise correlation is -1 to 1, and is used to determine whether two random variables are 'correlated' (i.e. move in relation) to each other. 1 is perfect correlation, 0 is uncorrelated, -1 is perfect inverse correlation (move perfectly in opposite directions).

Comment (jerclark) [reply README.md]: Played around with data visualization:

pandas.plot(kind="bar/line/hist, etc...")

But the most interesting was the pandas.scatter_matrix(df)

Provides an entire pairwise matrix of histograms. Really displays some interesting possible correlation patterns.

Comment (jerclark) [reply README.md]: Did some work with STEP 6 preprocessing data, but I want to revisit to really get my head around it. I understand things like setting a range, setting the mean to 0 and std to 1, I need to understand a little more.
Did set a range, and understood the output. Still need to understand what it looks like to 'remove' the mean.

Comment (jerclark) [reply README.md]: Found a good article about standardizing and normalizing data:
http://www.dataminingblog.com/standardization-vs-normalization/

Normalizing is converting the range of a feature to [0,1]. The simple formula:
x_new = (x - x_min) / (x_max - x_min)
This will always result in a fraction between 0 and 1, where the denominator is the original scale and the numerator is the pre-scaled value.

Standardizing is the process of making zero the 'mean' and applying unit variance.
x_new = x - mean(x) / std(x) 

While normalizing will force the same scale for all variables [0 - 1], if there are outliers the 'normal' data will fit into a very narrow band. 

Standardizing data, however, remains 'unbounded' so while it makes the mean 0 for all variables and 'shrinks' the scale, variables with single extreme outliers could throw off the scaling.

Unit variance means that the variance is one, which will also make the stdev 1 (since stdev is the square root of variance)

Found this neat visualization of standard deviation (Actually, image depicts variance, but stdev is square root of variance) https://en.wikipedia.org/wiki/Standard_deviation#/media/File:Variance_visualisation.svg

Comment (jerclark) [reply README.md]: Next up: Resampling algorithms. Split training data...some to train, and some to validate.
I need to do some more research into the kfold situation.

Comment (jerclark) [reply README.md]: Spent more time on exercises. I think I got the abalone dataset to be cross validated by making the targets integers (age). When I included age in the original training/validation data, the accuracy of Lasso was up around 98%? WTF is lasso?

Comment (jerclark) [reply README.md]: My lasso results were bunk from yesterday. Interestingly, LogisticRegression with Kfold 10 only returns ~25% accuracy with the abalone data set, but much higher accuracy with the diabetes data set here: "https://goo.gl/vhm1eU". I think there are far fewer records.

Also notices that LeaveOneOut cross validation (which is the same as KFold(n_folds=1)) takes a bit longer to run.

Comment (jerclark) [reply README.md]: Read about cross validation here: http://scikit-learn.org/stable/modules/cross_validation.html

And it was helpful. KFold Basically break the data down into k-trials, where each trial splits the data into equal parts training and validation. Use LOO to produce n folds, where the training set is all samples but one.

Comment (jerclark) [reply README.md]: Read about log loss here: https://www.google.com/amp/s/www.r-bloggers.com/making-sense-of-logarithmic-loss/amp/?client=safari

Didn't fully digest the math, but basically higher log loss is worse...and low confidence (high-probability of the wrong classification) is penalized over medium confidence probability.

Comment (jerclark) [reply README.md]: LogLoss is negative, and the higher the number the worse. Accuracy scoring is positive, and appears to be a percentage of the number of accurately predicted samples.

Comment (jerclark) [reply README.md]: Confusion matrix:
A matrix Ci,j where i is the actual classification group and j is the predicted classification group:
For example

        | Cat | Dog | Bird
Cat   | 3 | 2 | 0
Dog | 1 | 4 | 0
Bird | 0 | 0 | 11

Anything outside the diagonal is an 'error'
Then there's the table of confusion, which is a binary analysis of a single class and describes accuracy in terms of true positives, false positives, true negatives, false negatives
True is 'good' false is 'bad'
For the dog above:

        T   |   F
P   |  4   | 2
      ---------
N   |  14  |  1


This was my code to generate a confusion matrix:
```
from pandas import read_csv
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
url = "https://goo.gl/vhm1eU"
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
dataframe = read_csv(url, names=names)
array = dataframe.values
X = array[:,0:8]
Y = array[:,8]
kfold = KFold(n_splits=10, random_state=7)
model = LogisticRegression()
model = model.fit(X, Y)
prediction = model.predict(X)
cm = confusion_matrix(Y, prediction)
```

Comment (jerclark) [reply README.md]: Next up: classification report
http://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html

Comment (jerclark) [reply README.md]: Created a basic classification report for the same dataset as above...
             precision    recall  f1-score   support

        0.0       0.79      0.90      0.84       500
        1.0       0.74      0.55      0.63       268

avg / total       0.77      0.77      0.77       768

A good description of the values is here:
http://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_fscore_support.html

Intuatively, precision is the ability of the model to 'not falsely classify a sample' (Did I call that cat over there a dog?), while recall is the ability to 'not miss proper classification of a sample' (How many dogs did I miss?). 

The precision is:

Comment (jerclark) [reply README.md]: Setup regression problem to predict mpg. Here's my sample code with a bunch of placeholder models. Not sure what the MSE actually means:

```
# Cross Validation Classification LogLoss
import numpy
from pandas import read_table
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
import sklearn.linear_model as lm
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.data"
names = ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'model year', 'origin', 'car name']
dataframe = read_table(url, names=names, delimiter="\s+", na_values="?")
dataframe = dataframe[(~numpy.isnan(dataframe['horsepower']))]
array = dataframe.values
X = array[:,1:8]
Y = array[:,0]
linregression = lm.LinearRegression()
ridge = lm.Ridge()
lasso = lm.Lasso()
lars = lm.Lars()
omp = lm.OrthogonalMatchingPursuit()
br = lm.BayesianRidge()
model = linregression
model = model.fit(X, Y)
prediction = model.predict(X)
r2 = r2_score(Y, prediction)
mse = mean_squared_error(Y, prediction)
print("R2: %.3f, MSE: %.3f") % (r2, mse)
```

Comment (jerclark) [reply README.md]: MSE - mean squared error. The lower the better. The value is the average of the squares of the 'errors' (or, difference between the predicted and actual).

Comment (jerclark) [reply README.md]: Algorithm Tuning (parameter tuning?) Started looking at this. Ran the sample code, and it works. Looked a little bit into the estimator interface and exactly what GridSearchCV does. Came across this link: http://scikit-learn.org/stable/tutorial/machine_learning_map/
Helpful guide to pick an estimator. 

Next step is to try different parameters on the ridge function to see what I get.

Comment (jerclark) [reply README.md]: OK - algorithm tuning. GridSearchCV will take a 'grid' of parameters (think matrix master), where a sample of values is specified for each parameter being tested. The CV will then test and score every combination, and make the best combination available in the "best_estimator". 
As was noted in the example, I only tried different alpha versions on GradientBoostingRegressor, but I didn't get dramatically better results when I used the "best-alpha" maybe that's. In fact, it seemed worse in my "spot-check" file than when using the defaults.

Didn't do any work with http://scikit-learn.org/stable/modules/grid_search.html#randomized-parameter-optimization, but I think I get the gist.
Rather than explicitly defining a finite grid, provide ranges or discrete sets and a number of iterations to run through - and it will randomly test different combinations.

**Comment by jerclark** [comment]
Saved a model. Run fit, then dump the model object to a binary file using pickle.dump.
Then use pickle.load(open(filename, "rb")) to re-load the model.

File: README.md (hunk 1 above)
Comment (jerclark) [reply README.md]: Started End-To-End hello world stuff. Using iris data set, revisiting plots, stats, and cross validation with Kfolds. Looks like SVC is the best so far - I could try others.
